client_test.go

```diff
@@ -939,38 +1033,395 @@ func TestCommandsSymClient(t *testing.T) {
+	}
+
+	setC2KeyCmd, err := CmdSetC2Key(newC2PubKey)
+	if err != nil {
+		t.Fatalf("failed to create SetC2Key command: %v", err)
 	}
 
+	badProtectedSetC2KeyCmd, err := e4crypto.ProtectSymKey(append(setC2KeyCmd, 0x01), newSharedKey)
+	if err != nil {
+		t.Fatalf("Failed to protect command: %v", err)
+	}
+	if _, err := c.Unprotect(badProtectedSetC2KeyCmd, receivingTopic); err == nil {
+		t.Fatal("Expected an error with a bad setC2Key Command length")
+	}
+
+	protectedSetC2KeyCmd, err := e4crypto.ProtectSymKey(setC2KeyCmd, newSharedKey)
+	if err != nil {
+		t.Fatalf("Failed to protect command: %v", err)
+	}
+
+	_, err = c.Unprotect(protectedSetC2KeyCmd, receivingTopic)
+	if err != nil {
+		t.Fatalf("Failed to unprotect SetC2Key command: %v", err)
+	}
+
+	// Unprotecting should now fail since C2 pub key has changed
+	_, err = c.Unprotect(protectedSetC2KeyCmd, receivingTopic)
+	if err == nil {
+		t.Fatal("Expected unprotect to fail with a new C2 key")
+	}
+
+	newSharedKey, err = curve25519.X25519(newC2PrivKey, e4crypto.PublicEd25519KeyToCurve25519(newClientPubKey))
+	if err != nil {
+		t.Fatalf("curve25519 X25519 failed: %v", err)
+	}
+	newSharedKey = e4crypto.Sha3Sum256(newSharedKey)
+
+	// Unknown command
+	unknownCmd := []byte{0xFF}
+	protectedUnknownCmd, err := e4crypto.ProtectSymKey(unknownCmd, newSharedKey)
+	if err != nil {
+		t.Fatalf("Failed to protect command: %v", err)
+	}
+
+	_, err = c.Unprotect(protectedUnknownCmd, receivingTopic)
+	if err != ErrInvalidCommand {
+		t.Fatalf("Invalid error when unprotecting command: got %v, wanted %v", err, ErrInvalidCommand)
+	}
+}
+
+func assertClientTopicKey(t *testing.T, exists bool, c Client, topicHash []byte, topicKey []byte) {
+	tc, ok := c.(*client)
+	if !ok {
+		t.Fatalf("Unexpected type: got %T, wanted client", c)
+	}
+
+	k, ok := tc.TopicKeys[hex.EncodeToString(topicHash)]
 	if exists {
-		if !bytes.Equal(k, topicKey) {
+		if !ok {
+			t.Fatalf("Expected client to have topic %s key", hex.EncodeToString(topicHash))
+		} else if !bytes.Equal(k, topicKey) {
 			t.Fatalf("Invalid topic key: got %v, wanted %v", k, topicKey)
 		}
+	} else if ok {
+		t.Fatalf("Expected client to not have topic %s key", hex.EncodeToString(topicHash))
 	}
 }
 
-func assertContainsPubKey(t *testing.T, c Client, id []byte, key []byte) {
+func assertClientPubKey(t *testing.T, exists bool, c Client, id []byte, key []byte) {
 	pks, err := c.getPubKeys()
 	if err != nil {
 		t.Fatalf("Failed to retrieve pubkeys: %v", err)
 	}
 
 	pk, ok := pks[hex.EncodeToString(id)]
-	if !ok {
-		t.Fatal("Expected pubkey to be set on client")
-	}
-	if !bytes.Equal(pk, key) {
-		t.Fatalf("Invalid pubKey: got %v, wanted %v", pk, key)
+	if exists {
+		if !ok {
+			t.Fatalf("Expected pubkey for id %v to be set on client", id)
+		} else if !bytes.Equal(pk, key) {
+			t.Fatalf("Invalid pubKey: got %v, wanted %v", pk, key)
+		}
+	} else if ok {
+		t.Fatalf("Expected client not having pubkey for id %s", id)
```

Comment:
Thanks for the comprehensive test but I think that we should break it out into various tests of their own and use a common infrastructure plus tear down
```go
func setupForPubclientTests(t *testing.T) (*Client, ....) {
        clientID := e4crypto.HashIDAlias("client1")
	clientPubKey, clientSKey, err := ed25519.GenerateKey(nil)
	if err != nil {
		t.Fatalf("failed to generate ed25519 keys: %v", err)
	}

	c2PrivKey := e4crypto.RandomKey()
	c2PubKey, err := curve25519.X25519(c2PrivKey, curve25519.Basepoint)
	if err != nil {
		t.Fatalf("failed to generate curve public key: %v", err)
	}

	topic := "topic1"
	topicHash := e4crypto.HashTopic(topic)

	receivingTopic := TopicForID(clientID)

	sharedKey, err := curve25519.X25519(c2PrivKey, e4crypto.PublicEd25519KeyToCurve25519(clientPubKey))
	if err != nil {
		t.Fatalf("curve25519 X25519 failed: %v", err)
	}
	sharedKey = e4crypto.Sha3Sum256(sharedKey)

	c, err := NewClient(&PubIDAndKey{
		ID:       clientID,
		Key:      clientSKey,
		C2PubKey: c2PubKey,
	}, NewInMemoryStore(nil))
	if err != nil {
		t.Fatalf("Failed to create client: %v", err)
	}
        return client, ...
}

func TestCommandsPubClient_badSetTopic(t *testing.T) {
        client, ... := setupForPubclientTests(t)
	badProtectedSetTopicCmd, err := e4crypto.ProtectSymKey(append(setTopicCmd, 0x01), sharedKey)
	if err != nil {
		t.Fatalf("Failed to protect command: %v", err)
	}
	if _, err := client.Unprotect(badProtectedSetTopicCmd, receivingTopic); err == nil {
		t.Fatal("Expected an error with a bad setTopic Command length")
	}
}
...
Many more sub cases that are isolatable here.
```
where `...` means you can add more arguments